### PR TITLE
HPCC-24060 Add a DaemonSet to prepull HPCC image to k8s nodes

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Tested on Docker Desktop where it has no ill-effect but hard to say if it has any positive imact either.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith What do you think?

**Comment by jakesmith** [comment]
@richardkchapman - Interesting, I can see it does help the case for the 1st run of an installation where there are unused nodes in a node pool. I don't think it helps where auto-scaling though, because the auto-scaling would allocate new nodes due to pod resource pressure and the DaemonSet wouldn't get to pull the image any more quickly than the pod would have done.
And I guess you'd want to leave this option off, if you had a K8s cluster and node pool(s), that were being used for HPCC + other things.

Side question, how does docker handle 2 requests to pull an image? (which could happen anyway with 2 pods running on a clean node). I guess it must handle concurrency efficiently already, e.g. spot that there's an existing layer being pulled and block.

I see others are doing the same, for similar reasons.
Came across this : https://codefresh.io/kubernetes-tutorial/single-use-daemonset-pattern-pre-pulling-images-kubernetes/

Should we avoid starting the our container (and using a bit of OS resources), and run a minimal one instead and issues a docker pull <image> ?
Should we use an initContainer pattern, to avoid the DaemonSet restartPolicy potentially kicking in?

**Comment by richardkchapman** [comment]
That article was the inspiration for this PR, and the comment at the bottom was why I did it this way rather than the way that the article described.

I think the ideas is that if nodes are held in reserve ready for pods, then this would help them be "more ready" - i.e it will move the bulk of the image pull time from the "start the first pod on a node" to the "start the node" time. Whether that makes a difference in practice, I don't know.

**Comment by jakesmith** [comment]
right, ok, I missed that comment.
Makes sense not to use docker pull .. could still use an initContainer to avoid the DaemonSet restartPolicy ?
And the init container could finish, and hence not be running base (e.g. ubuntu20.04) image and any  resources it might use.
The running container (pause in the articles example) could be busybox or something trivial.

**Comment by richardkchapman** [comment]
Probably not worth merging until we are able to test on a system where something like this would make a difference.

**Comment by richardkchapman** [comment]
@mckellyln @jakesmith Are we in a position to do any more testing on this?

**Comment by jakesmith** [comment]
There are a couple of further comments in the article now (https://codefresh.io/kubernetes-tutorial/single-use-daemonset-pattern-pre-pulling-images-kubernetes/). Last one mentions auto scaling.

The daemonset pod is only going to be deployed once the node pool scales, and isn't it only ever going to scale because resources can't be satisfied and therefore trigger scaling to run the pod that required it.
I'm not sure the daemonset helps in that case.

If there was a non-scaling pool of nodes in V8s that was always running, then it would help get the image to the nodes that weren't otherwise being used in advance.

@richardkchapman - I could test this on Azure with and without an auto scaling node pool.
I don't think I'll be able to tell if it helps in the auto-scaling case or not though..

File: helm/hpcc/templates/daemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      name: prepull 
+  template:
+    metadata:
+      labels:
+        name: prepull 
+    spec:
+      containers:
+      - name: prepull 
+{{ include "hpcc.addImageAttrs" (dict "root" $ "me" . "imagename" "platform-core") | indent 8 }}
+        command: ["sleep", "10000d"]
```
Comment (ghalliday) [inline helm/hpcc/templates/daemonset.yaml:18]: you'll have retired before it times out?
Will this continue to poll to see if there are new images?

How does this approach compare with https://gist.github.com/itaysk/7bc3e56d69c4d72a549286d98fd557dd

**Comment by richardkchapman** [comment]
Looking at the side @ghalliday linked to, it looks like they are taking a slightly different approach, at least in so far as they provide a helm chart that can be added to any installation to prepull any image. Perhaps that's a better approach than building prepull capability into HPCC.

I think I will close this on the basis of insufficient (proven) utility versus one more thing to maintain/document.

**Pull request closed without merging**